modules/gateway/peersmanager.go

```diff
@@ -73,71 +72,97 @@ func (g *Gateway) permanentPeerManager(closedChan chan struct{}) {
-			defer func() {
-				<-connectionLimiterChan
-			}()
-
-			if err := g.threads.Add(); err != nil {
+			g.log.Debugln("[PPM] Fetched a random node:", addr)
+
+			// We need at least some of our outbound peers to be remote peers. If
+			// we already have reached a certain threshold of outbound peers and
+			// this peer is a local peer, do not consider it for an outbound peer.
+			// Sleep briefly to prevent the gateway from hogging the CPU if all
+			// peers are local.
+			if numOutboundPeers >= maxLocalOutboundPeers && addr.IsLocal() && build.Release != "testing" {
+				g.log.Debugln("[PPM] Ignorning selected peer; this peer is local and we already have multiple outbound peers:", addr)
+				if !g.managedSleep(unwantedLocalPeerDelay) {
+					return
+				}
+				continue
+			}
+
+			// Try connecting to that peer in a goroutine. Do not block unless
+			// there are currently 3 or more peer connection attempts open at once.
+			// Before spawning the thread, make sure that there is enough room by
+			// throwing a struct into the buffered channel.
+			g.log.Debugln("[PPM] Trying to connect to a node:", addr)
+			connectionLimiterChan <- struct{}{}
+			go func(addr modules.NetAddress) {
+				// After completion, take the struct out of the channel so that the
+				// next thread may proceed.
+				defer func() {
+					<-connectionLimiterChan
+				}()
+
+				if err := g.threads.Add(); err != nil {
+					return
+				}
+				defer g.threads.Done()
+				// peerManagerConnect will handle all of its own logging.
+				g.managedPeerManagerConnect(addr)
+			}(addr)
+
+			// Wait a bit before trying the next peer. The peer connections are
+			// non-blocking, so they should be spaced out to avoid spinning up an
+			// uncontrolled number of threads and therefore peer connections.
+			if !g.managedSleep(acquiringPeersDelay) {
 				return
 			}
-			defer g.threads.Done()
-			// peerManagerConnect will handle all of its own logging.
-			g.managedPeerManagerConnect(addr)
-		}(addr)
-
-		// Wait a bit before trying the next peer. The peer connections are
-		// non-blocking, so they should be spaced out to avoid spinning up an
-		// uncontrolled number of threads and therefore peer connections.
-		if !g.managedSleep(acquiringPeersDelay) {
-			return
 		}
 	}
 }
+
+// buildPeerManagerNodeList returns the gateway's node list in the order that
+// permanentPeerManager should attempt to connect to them.
+func (g *Gateway) buildPeerManagerNodeList() []modules.NetAddress {
+	// flatten the node map, inserting in random order
+	nodes := make([]modules.NetAddress, len(g.nodes))
+	perm := fastrand.Perm(len(nodes))
+	for _, node := range g.nodes {
+		nodes[perm[0]] = node.NetAddress
+		perm = perm[1:]
+	}
+
+	// swap the outbound nodes to the front of the list
+	numOutbound := 0
+	for i, node := range nodes {
+		if g.nodes[node].WasOutboundPeer {
+			nodes[numOutbound], nodes[i] = nodes[i], nodes[numOutbound]
+			numOutbound++
+		}
+	}
+	return nodes
```

Comment:
Added in https://github.com/NebulousLabs/Sia/commit/4a09c3770ca518a9bc3032ed4a7e2e0b64b6b4c8